### DISCUSSION TITLE
Reflection.Emit and the dynamic keyword

### DISCUSSION THREAD

**Discussion by davepruitt** [dotnet/maui#12397] [Q&A]
I am just curious what is supported with regard to using C#'s `dynamic` keyword and/or Reflection.Emit in .NET Maui apps? I am currently working on an app that I plan to release on both iOS and Android. 

As an example simple use case in my app: sometimes I am given two variables that I need to add together. However, due to intricacies in the code, sometimes those variables will be handed to me as type `object`. Obviously if you try to do an add operation on two variables of type `object`, the code will give you a compile-time error:
```
//Of course this will not compile
public object simple_example_add (object a, object b)
{
    return a + b;
}
```

One solution to this is to use dynamics. So something like this would compile and work:
```
public object simple_example_add (object a, object b)
{
    dynamic result = (dynamic)a + (dynamic)b;
    return result;
}
```

This would not work on Xamarin Forms. I believe it either failed to compile in Xamarin Forms, or it just crashed any time I tried it.

My current workaround is to use a Lua interpreter (NLua), which handles it well:
```
public object simple_example_add (object a, object b)
{
    using (NLua.Lua state = new NLua.Lua())
    {
        state["x"] = a;
        state["y"] = b;
        var result = state.DoString("return x + y")[0];
        return result;
    }
}
```

But it seems silly to have to use a Lua interpreter just to be able to do something like this.

Anyways, what's the word on `dynamic` support in .NET Maui for each platform?

**Comment by AnthonyIacono** [comment]
> I am just curious what is supported with regard to using C#'s `dynamic` keyword and/or Reflection.Emit in .NET Maui apps? I am currently working on an app that I plan to release on both iOS and Android. 
> 
> 
> 
> As an example simple use case in my app: sometimes I am given two variables that I need to add together. However, due to intricacies in the code, sometimes those variables will be handed to me as type `object`. Obviously if you try to do an add operation on two variables of type `object`, the code will give you a compile-time error:
> 
> ```
> 
> //Of course this will not compile
> 
> public object simple_example_add (object a, object b)
> 
> {
> 
>     return a + b;
> 
> }
> 
> ```
> 
> 
> 
> One solution to this is to use dynamics. So something like this would compile and work:
> 
> ```
> 
> public object simple_example_add (object a, object b)
> 
> {
> 
>     dynamic result = (dynamic)a + (dynamic)b;
> 
>     return result;
> 
> }
> 
> ```
> 
> 
> 
> This would not work on Xamarin Forms. I believe it either failed to compile in Xamarin Forms, or it just crashed any time I tried it.
> 
> 
> 
> My current workaround is to use a Lua interpreter (NLua), which handles it well:
> 
> ```
> 
> public object simple_example_add (object a, object b)
> 
> {
> 
>     using (NLua.Lua state = new NLua.Lua())
> 
>     {
> 
>         state["x"] = a;
> 
>         state["y"] = b;
> 
>         var result = state.DoString("return x + y")[0];
> 
>         return result;
> 
>     }
> 
> }
> 
> ```
> 
> 
> 
> But it seems silly to have to use a Lua interpreter just to be able to do something like this.
> 
> 
> 
> Anyways, what's the word on `dynamic` support in .NET Maui for each platform?

Just an FYI, dynamic code generation is not allowed on iOS at all.

**Reply by davepruitt** [reply]
And yet there are hundred, if not thousands, of apps that allow users to write code on iOS, and those apps even interpret and/or compile/run the code _on the device_. So, yes, there is "dynamic code generation" in iOS apps.

**Reply by AnthonyIacono** [reply]
Although the policy may have changed, I was under the impression that JIT was prohibited.

EDIT: looks like iOS 14.2 introduced JIT, so I guess it might work to use Reflection.Emit

**Reply by davepruitt** [reply]
Just a few examples of iOS apps that allow users to author code (and then execute that code) directly on iOS devices (code is not compiled in the cloud):

1. Codea
2. LuaStudio
3. NLuaBox
4. LowRes NX
5. Brace Editor
6. DraftCode Offline PHP IDE
7. Touch Lua
8. Sketch Lua
9. Nano Basic Interpreter
10. Nano C language interpreter
11. GG Basic

**Reply by AnthonyIacono** [reply]
> Just a few examples of iOS apps that allow users to author code (and then execute that code) directly on iOS devices (code is not compiled in the cloud):
> 
> 
> 
> 1. Codea
> 
> 2. LuaStudio
> 
> 3. NLuaBox
> 
> 4. LowRes NX
> 
> 5. Brace Editor
> 
> 6. DraftCode Offline PHP IDE
> 
> 7. Touch Lua
> 
> 8. Sketch Lua
> 
> 9. Nano Basic Interpreter
> 
> 10. Nano C language interpreter
> 
> 11. GG Basic
> 
> 
> 
> 

Right on. Yeah looks like iOS 14.2 started allowing JIT with the right entitlement. For what it is worth, the apps you listed are mostly interpreters. Some might be compiled, but interpreted code is much different than compiled in terms of security policy.

That being said, Reflection.Emit should work with the JIT entitlement

**Reply by hartez** [reply]
Historically it _wasn't_ an option, but that's changed and there's work being done to enable it: https://devblogs.microsoft.com/xamarin/introducing-xamarin-ios-interpreter/
